### ISSUE TITLE
`replacedBy` not affecting static methods

### ISSUE THREAD

**Issue by Piinks** [dart-lang/sdk#50698]
Working on a change in flutter/flutter, I deprecated a static method in favor of a new method.

I've tried this a couple of ways now, but the transform has never worked:

I deprecated

```dart
@Deprecated('Use preferredSizeFor...')
static double preferredHeightFor(BuildContext context, Size preferredSize) {...}

/// Replacement
Size preferredSizeFor(BuildContext context) {...}
```

with this transform

```yaml
  # Changes made in TBD
  - title: "Migrate to 'preferredSizeFor'"
    date: 2022-12-12
    element:
      uris: [' material.dart' ]
      method: 'preferredHeightFor'
      inClass: 'AppBar'
    changes:
    - kind: 'replacedBy'
      newElement:
        uris: ['material.dart']
        method: 'preferredSizeFor'
        inClass: 'AppBar'
    - kind: 'removeParameter'
      index: 1
```

And I guess I expected this change:

```dart
// Before
AppBar.preferredHeightFor(context, preferredSize);

// After
preferredSizeFor(context);
```

But nothing happened.

The new method is not static, so I know this is not a complete transformation, I was hoping that I could at least get users in the right direction. Is this possible? Or is there another way?

**Comment by bwilkerson** [comment]
According to the documentation, `replacedBy` doesn't currently support replacing a static method with an instance method:

> There isn’t currently any way for the tool to detect whether a method, getter, or setter is static, so it’s possible to specify an instance member. As a result, specifying an instance member (as either the replaced or replacing element) might produce invalid changes for your users.

I suspect that there are two reasons for that. The first is completely pragmatic: we didn't have a use case for that kind of transform, so we didn't implement it. Now we have a use case.

The second is that we hadn't really considered supporting partial replacements. I'm of two minds about that. It can sometimes be helpful to make the parts of the change that we're able to make in order to reduce the work left for the user. But what we need to be careful of is that the remaining set of changes are easier to understand and make than the original set of changes would have been.

For example, in this case would it be easier for users to be told that `preferredHeightFor` is deprecated with the ability to find the deprecation notice that explains how to convert their code, or to be told that `preferredSizeFor` isn't defined, with no way to figure out that they need to add a target (and presumably a `.height` after the invocation)? After `preferredHeightFor` is removed, then the user won't be able to navigate to the deprecation message, so some of the advantage is gone, but it still might be easier to search for `AppBar.preferredHeightFor` than `preferredSizeFor` in order to figure out what to do about it.

**Comment by bwilkerson** [comment]
I gave this a low priority because it's still under discussion, not because supporting Flutter isn't important.

**Comment by Piinks** [comment]
> I gave this a low priority because it's still under discussion, not because supporting Flutter isn't important.

Oh no worries at all. :) 

I agree, partial fixes are a murky area. I don't know if it is worth pursuing support for a partial fix like this, I was mainly curious if any existing conventions could help. I am ok with closing this issue, or waiitng to see if another use case comes along.

> For example, in this case would it be easier for users to be told that preferredHeightFor is deprecated with the ability to find the deprecation notice that explains how to convert their code, or to be told that preferredSizeFor isn't defined, with no way to figure out that they need to add a target (and presumably a .height after the invocation)?

I totally agree. The partial fix in this scenario seems like it would cause more harm than help.

**Comment by asashour** [comment]
I guess the analyzer can know if the method/getter is static or not (`ExecutableElement`.`isStatic`).

As a user, I would expect an instance method to be replaced by a static one, since it is straight-forward.


Replacing a static method with an instance one, is mainly about the instance variable to add, in which case there could be a possibility to specify that variable (which must be existing in the caller code, and can also be `this`).

Changing the parameters or return type of the method is another type of `change` I guess.

**Comment by bwilkerson** [comment]
> I guess the analyzer can know if the method/getter is static or not ...

That information isn't currently available after the old member has been removed.

> As a user, I would expect an instance method to be replaced by a static one, since it is straight-forward.

Given `expression.instanceMethod(arguments)`, we could produce `ClassName.staticMethod(arguments)`, but what do we do with `expression`? Some expressions have side-effects, others don't, but we can't do enough analysis to know which case we're in with any certainty. We shouldn't delete it if it has a side-effect, and shouldn't leave it if it doesn't.

> .... there could be a possibility to specify that variable ...

Do you mean "specify it in the data file", or "allow the user to specify it" (or something else)? I'm not sure either is a viable option, but I'm curious.

> Another option is to have any string which could possibly be an expression, but the validation is lost in this case.

Can you provide an example? I don't know what you're thinking of here.

**Comment by asashour** [comment]
> That information isn't currently available after the old member has been removed.

I didn't think about that, because most of the scenarios of the test cases, the old element still exists (and deprecated). But if it is removed, it won't be there.

> Some expressions have side-effects, others don't, but we can't do enough analysis to know which case we're in with any certainty. We shouldn't delete it if it has a side-effect, and shouldn't leave it if it doesn't.

Some cases (e.g. local variables), don't have side effects, so they can be removed. Regarding the expressions, it won't harm if they are left even if they don't have side-effects, and leave this to (or include the logic of) the lints. As always, the analyzer deletes code when it is sure it doesn't affect the semantics, otherwise leaves it compilable.


> > .... there could be a possibility to specify that variable ...

Sorry for not being clear enough. A user is part of Flutter team who writes data-drive fixes, an end-user is a developer building web application using Flutter.

`ClassName.staticMethod(argument)` needs to be `expression.instanceMethod(arguments)`, so the user (Flutter team) can specify a string in the transform `change` for example:

```yaml
...
  changes:
    - kind: 'replacedBy'
      newElement:
        method: 'instanceMethod'
        inClass: 'newInstanceClass'
        variable: 'see below'
```
And the `variable` can be one of:
   - `this`
   - `variableName` or
   - `variableName2.anotherMethod()`

So the transform does the following respectively:
  - If the class which contains the code calling `ClassName.staticMethod()` is a subclass of the change element `newInstanceClass`, it will be changed to `instanceMethod()`
  - If the `variableName` is a local variable in the method/function which calls `ClassName.staticMethod()`, it is changed to `variableName.instanceMethod(argument)`
  - If the `variableName2` is a local variable/field in the method/function which calls `ClassName.staticMethod()`, it is changed to `variableName2.anotherMethod().instanceMethod(arguments)`

> > Another option is to have any string which could possibly be an expression, but the validation is lost in this case.

What was meant, that `variable` can be a free text, written by Flutter team, and there is no validation so:

  - `ClassName.staticMethod()`, it is changed to `<variable text>.instanceMethod(arguments)`

They are all suggestions, to be confirmed by use cases, in the Flutter migration journey.

**Comment by bwilkerson** [comment]
> They are all suggestions, to be confirmed by use cases, in the Flutter migration journey.

I agree that we should have concrete use cases before attempting to implement this kind of transformation, but I remain skeptical that we'll be able to do a good enough job that doing so would actually bringing value to users.

**Comment by keertip** [comment]
Here is a concrete use case [link](https://github.com/flutter/flutter/pull/117480/files) for replacing static methods. One more [example](https://github.com/flutter/flutter/pull/117486/files). 

Some more details on this use case 

```
if context available
   if accessing functionality from platformDispatcher
      replace WidgetsBinding.instance.window with View.of(context).platformDispatcher
   else 
      replace WidgetsBinding.instance.window with View.of(context)
else no context available
   if accessing functionality from platformDispatcher
      replace window with platformDispatcher 

```

//cc @goderbauer
